src/reporters/reporter-util.ts

```diff
@@ -0,0 +1,209 @@
+  }
+
+  return {
+    file,
+    line: frame.line || 0,
+    column: frame.column || 0,
+    function: frame.function,
+  };
+}
+
+function filterLibInternals(file: string) {
+  // To signore filtering the stack trace on tests
+  if (process.env.TEST_OVERRIDE) {
+    return true;
+  }
+  if (file.startsWith(SYNTHETICS_PATH)) return false;
+  // should have been filtered by node_modules, but just in case
+  if (file.startsWith(PW_CORE_PATH)) return false;
+  return true;
+}
+
+function constructStackFromFrames(frames: StackFrame[]) {
+  const stackLines: string[] = [];
+  for (const frame of frames) {
+    stackLines.push(
+      `    at ${frame.function ? frame.function + ' ' : ''}(${frame.file}:${
+        frame.line
+      }:${frame.column})`
+    );
+  }
+  return stackLines;
+}
+
+export function prepareError(error: Error) {
+  const stack = error.stack;
+  const lines = stack.split('\n');
+  let startAt = lines.findIndex(line => line.startsWith('    at '));
+  if (startAt === -1) startAt = lines.length;
+  const message = lines.slice(0, startAt).join('\n');
+  const stackLines = lines.slice(startAt);
+  // figure out the location of the journey/step that throws the error and
+  // correct stack line corresponding to that error
+  const stackFrames: StackFrame[] = [];
+  for (const line of stackLines) {
+    const frame = prepareStackFrame(line);
+    if (!frame || !frame.file) continue;
+    stackFrames.push(frame);
+  }
+
+  let location: Location | undefined;
+  if (stackFrames.length) {
+    const frame = stackFrames[0];
+    location = {
+      file: frame.file,
+      column: frame.column || 0,
+      line: frame.line || 0,
+    };
+  }
+
+  return {
+    message,
+    stack: constructStackFromFrames(stackFrames).join('\n'),
+    location,
+  };
+}
+
+export function highLightSource(location: Location): string {
+  if (!location) {
+    return;
+  }
+
+  try {
+    const source = readFileSync(location.file, 'utf-8');
+    const code = codeFrameColumns(
+      source,
+      { start: location },
+      { highlightCode: true }
+    );
+    return code;
+  } catch (_) {
+    // ignore error
+  }
+}
+
+// serializeError prefers receiving proper Errors, but since at runtime
+// non Error exceptions can be thrown, it tolerates though. The
+// redundant type Error | any expresses that.
+export function serializeError(
+  error: Error | any,
+  highlightCode = true
+): TestError {
+  if (!(error instanceof Error) || !error.stack) {
+    return { message: `thrown: ${inspect(error)}` };
+  }
+
+  const testErr: TestError = { message: error.message };
+  const { message, stack, location } = prepareError(error);
+  testErr.message = message;
+  if (stack) {
+    testErr.stack = stack;
+  }
+  if (location && highlightCode) {
+    testErr.location = location;
+    testErr.source = highLightSource(testErr.location);
+  }
+  return testErr;
+}
+
+export function renderError(error: TestError) {
+  const summary = [];
+  summary.push('');
```

Comment:
nit: what do you think about adding a comment explaining why we add `''`?

Reply:
Will do.